### ISSUE TITLE
Make docs site downloadable / set up a template-based SPA option

### ISSUE THREAD

**Issue by bob2517** [Active-CSS/active-css#129]
Add a button to download a zip containing a searchable docs site. All the files are statically generated anyway so it's not going to be much work to implement. Make an auto-generated JavaScript file that runs with the offline website grunt that contains the map of keywords and the search functionality. In the config, instead of using ajax, run the function instead. Just use a global window variable for storing the search object, or local storage, and the function can just return straight html to render..

**Comment by bob2517** [comment]
Could have an archive of older versions. Starting with the latest one on release. That way I don't have to maintain multiple versions live and it can stay simple. People worried about version changes can just download the matching archive zip.

Give it it's own page in the downloads section of the docs site.

**Comment by bob2517** [comment]
I'm going to set up a simple repo on github for downloadable versions of the docs. I think that's probably the most sensible way to do it. That way if the docs site disappears for any reason then there's still a stable place to get at the docs. I'm not put the whole docs on github as browseable online files, as it's going to be double work for me and not practical for my one-man setup. It's obviously also completely pointless as the docs site is faster anyway and people can just go there for online browsing.

**Comment by bob2517** [comment]
With the switch over to Action CSS in mind, I'm going to put it under the main action css org area and just have separate dirs on there for Active CSS and later on, Action CSS. Something like that.

**Comment by bob2517** [comment]
Or not. May just use an active CSS repo and put things there and do a new repo containing the old active css files. I dunno. Doesn't matter particularly at this point.

**Comment by bob2517** [comment]
There must be a way for a local website to be able to run local ajax commands to the same directory the local website is running on.

There are zero security implications for that behaviour in the website is sandboxed into that directory. Otherwise it makes it impossible to run a local website, using just HTML, CSS and JavaScript, without making a Cordova app or something equally over-engineered for a basic local filesystem. Or does Google want devs to only build phone apps tied to Android and are basically calling the shots - starting with disabling local file access?

What am I supposed to do, write a desktop web app? I can't believe that's the only solution. WIll look into it some more.

**Comment by bob2517** [comment]
Hmmm... from the look of it I might have to build an non-fancy version of the docs site for offline use due to limitations put on browsers for using local files. It would have been nice to be able to use ajax for local files, but I can see why it's been turned off in the browser. The search box would have to be fully JS, but then I was going to do that anyway. Will try a few things out.

**Comment by bob2517** [comment]
All the config needs to be inline - it's a good job I built inline ACSS into the core... I can see why people don't do this sort of thing much. The code editor's going to be interesting. Thinking aloud, I could give the code editor an option to take in an inline template, and split that up into HTML, CSS and ACSS. Any extra JS I can just include in the ACSS for the code editor. Would actually be quite easy. Extract any style tags, remove them, extract any acss-style tags, remove them, and the remainder is the HTML. That would be a neat thing to do and a bit radical - a fully-functional offline code editor. Wouldn't have to rely on potentially deprecating functionality from the browser either.

**Comment by bob2517** [comment]
Now unsure if I can use github to store zip files. Need to check it out and will come up with an alternative plan if not. Worst case scenario I'll use my live server, but I'd rather not.

**Comment by bob2517** [comment]
Was thinking - how big would data take up for a docs search box? It shouldn't be more than 20kb, just for keywords and descriptions. If I can get the offline search box working, I could make the live site fully static and cacheable. That might be slightly revolutionary - not seen that recommended anywhere. I don't see anyone doing it live, but someone must have done it.

**Comment by bob2517** [comment]
Ok - back on this at last! Target: end of the weekend, or thereabouts.

**Comment by bob2517** [comment]
As I can't run ajax commands locally, I'm looking at putting each page into a template tag and just converting the whole site into one file, and then on nav, hiding and showing template contents depending on the page. HTML files aren't that big. At worst, the overall file size of the whole is going to be 250kb - I actually have no idea, but it shouldn't be more than that and probably less. Template tag inner stuff won't render, so page load should be fast enough. But nav would be very quick. Other stuff like css and images can be in different places. But all the HTML and example's code editor ACSS can just be in template tags and rendered when needed. Hmmm... need to check that ACSS in template tags doesn't currently load into current config. It shouldn't do, but will run into it shortly if there is a problem.

**Comment by bob2517** [comment]
Switching to doing a grunt ACSS minifier for this to try and keep the size as low as possible. Would be good to get it done anyway, but I actually have a reason to do it now.

**Comment by bob2517** [comment]
Actually I don't want the minifier for what I need to do, but I'll do a version of it anyway - probably keep it a private repo until I'm happy it's going to do something useful.

**Comment by bob2517** [comment]
Update on this issue: It works as an template-based (rather than ajax) single-file SPA offline, but in order to replicate the live site exactly it still needs the following:

1. Code editor support;
2. "Double-hash" support, like another hash within the new SPA modal documentation beyond the file name hash;
3. Offline search;
4. (optional) Get the support form working offline, always going via temporary local storage until network connection so that submittance can be callback verified (will need the proposed PWA upgrades to work properly, so this is optional and will probably not happen for the 2.5.1 release).

**Comment by bob2517** [comment]
The double-hash support is going to be implemented as multi-hash support. The URL will basically get hash splitted. Every appropriate hash item will get it's own hash trigger which will fire if it wasn't already in the URL once the underlying page is drawn. I need double-hash support for the offline docs site anyway and it would make sense to build in multi-level at this point rather than later.

Will probably need an add-hash and a remove-hash command which can target specific hash strings.

It sounds like a major upgrade but it shouldn't take that long to implement and I need it now anyway for this issue.

**Comment by bob2517** [comment]
Implementing multi-hash support instead of double-hash support. It's just as easy and a bit cleaner to add more functionality rather than less for this. This will need documenting.

**Comment by bob2517** [comment]
Multi-level hash support is now done offline. Will probably commit tomorrow - just want to spend a day with it to make sure all good. I needed this for the offline docs site, which has pages like this:

```
file:///K:/random-stuff/acss-offline/index.html#/manual/spa-image-modal.html#horse
```

Note the double-hashness.

All pages in the standalone offline docs use the file index.html and all inner pages are based on hash links which open and close template contents based on @pages, automatically for offline files that use #/, so it works like an SPA without changing anything. But also needed to have another hash for the modal popup to open and for the modal to popup when the page is refreshed, like any good SPA.

But multi-dimensional hashes are now supported which makes this more flexible, so more than two hashes can be added. As long as triggers are in \@pages per the other upgrade for 2.5.1 then it works.

For that there are two new commands:

```
add-hash: "#myHash" "a new title";    /* add-hash adds a hash if it isn't already there + optionally sets any new title. */
remove-hash: "#myHash" "an old title";    /* remove-hash remove a hash that is found + optionally sets any new title. */
```

**Comment by bob2517** [comment]
Note for tomorrow: Test manual a tags.

**Comment by bob2517** [comment]
Fixed point 3 above now: "Fix the case of inline ACSS not working when all the templates are loaded but it works on the first page."

It was due to the fact I was inserting template content into the page from within a custom command and not the render command. That was due to a limitation in the syntax for rendering templates. But of course it doesn't go through the core and so doesn't get correctly escaped, etc. like it would on a regular ajax load.

I've added a new command into the core, render-template-html, which is more flexible than what is there currently which only allows reference by tag ID. I wanted to store urls into a reference in the template so I could literally just display the template based on the url, which is stored in a data-ref attribute in the template tag. So after changing the url and title, etc. from regular SPA links on the page I just do the following instead of an ajax command and then use afterRenderTemplateHtml like I would afterAjax if I was using the ajax command for loading content:

```
#str-ContentInner {
	render-template-html: template[data-ref="{@href}"];
}
```

**Comment by bob2517** [comment]
Docs to do for the new command.

**Comment by bob2517** [comment]
Offline search is done. It's actually more accurate the the MySQL fulltext search so I should probably revisit that and work out what's going on with that. The technique I've used for this is to put each page section text into an array along with data in order to be able to show the search link for that page, then I just did a custom command when searching which splits each page by the search results to get a count of the appearing search term (there is probably a better way - will look at that), then put into another array of final results if it's there, and then finally sort the last array by number of appearances in descending order, then use the the live site component loop to display the results as I used the same array format at the ajax result on the live site. Works a little bit faster than the live site.

I could have used a more pure CSS version of searching by using the attributes, but I didn't want to bloat the DOM any more than it is already and wanted "number of appearances" search results anyway.

I could further work on this to prioritise by page name, but it works quite well and there probably isn't much point.

**Comment by bob2517** [comment]
It's pretty much done now. Just some more tests to do. The ajax command example doesn't work, but then you can't have ajax commands from a "file://" location, so there's not much I can do about that.

**Comment by bob2517** [comment]
Oh yeah - need to do the link to be able to download it as a zip. Will work on that tomorrow.

Then I think it's just a case of updating the docs and 2.5.1 can get released. There's a few more stability fixes that I need to fix for this one, which is good.

**Comment by bob2517** [comment]
I've set up a private repo (until release) that will contain the docs site in SPA form. Each core release will get a corresponding docs release which will get tagged as a specific release so it's easy to download. I've set up a docs download page on the docs website which will give direct download links to the specific zipped release on GitHub.

So marking this as done now. Just the docs site to update and that should be it for 2.5.1.
